### PR TITLE
[RecommendationSystemsTest] Adding hyperparameters to prevent data ga…

### PR DESCRIPTION
**Pull request by mlupon**
Summary:

Current RecommendationSystemsTest implementation data gates the final results, especially when using configurations with several bottom MLP layers. This is problematic because it can potentially hide bugs in the different phases of the test, as at some point the intermediate results are zero and then the final result returns the expected value--even if the implementation of the intermediate phases is wrong. 

Since finding a correct randomization of inputs that does not cause this effect for the infinite testing space is quite complex, we provide tools to the user to configure the inputs in such a way that prevents data-gating, generating intermediate tensor results different to zero. These hyperparameters are able to scale down the biases used in the MLP layers and shift the random value of the weights so they are not centered to zero. On top of that, we allow the user to explicitly force inputs to be negative. 

Usage:

  --act-factor=<number: Activation scale factor.
  --act-pos: Activation positive only.
  --bias-factor=<number>: Bias factor.
  --weight-offset=<number>: Weight offset.

Default behavior is the current behavior, which means that tensors are not patched. 

Example:

./tests/InterpreterRecommendationSystemTest  --gtest_filter="Interpreter/RecommendationSystemTest.RecSys_RWQuantized_SLWS_FP16/0"

zeroes the result, while the following run does not:

./tests/InterpreterRecommendationSystemTest  --bias-factor="1" --weight-offset="0.0010" --act-factor="1.0" --act-pos="true" --gtest_filter="Interpreter/RecommendationSystemTest.RecSys_RWQuantized_SLWS_FP16/0" 

Fixes:

https://github.com/pytorch/glow/issues/3012

Test Plan:

Running RecommendationSystemTest with different backends and configurations, dumping intermediate results to make sure that the intermediate results are not zero.

### CODE REVIEW

**Review by jfix71 — COMMENTED** [review]
Thanks for looking into this! Added a few comments.

File: tests/unittests/RecommendationSystemTest.cpp [hunk 1]
```diff
@@ -435,6 +458,41 @@ class RecommendationSystemTest : public BackendTest {
     }
   }
 
+  /// Transforms the content of a tensor based on the inputs
+  ///   div set to true, divides the content of the tensor by a scale factor,
+  ///   while set to false it shifts it.
+  ///   value is the factor which we want to divide or shift
+  ///   size is the size of the tensor
+  ///   internalType determines the precision of the tensor
```
[the remark this answers is not in the document]
Comment (mlupon) [reply tests/unittests/RecommendationSystemTest.cpp]: Fixed

File: tests/unittests/RecommendationSystemTest.cpp [hunk 2]
```diff
@@ -435,6 +458,41 @@ class RecommendationSystemTest : public BackendTest {
...
+        if (div) {
+          data.raw(i) = (float16_t)((float)data.raw(i) / value);
+        } else {
+          data.raw(i) = (float16_t)((float)data.raw(i) + value);
+        }
+      }
+      break;
+    }
+    default:
+      assert(false && "Can only patch floating-point tensors");
```
[the remark this answers is not in the document]
Comment (mlupon) [reply tests/unittests/RecommendationSystemTest.cpp]: Fixed

File: tests/unittests/RecommendationSystemTest.cpp [hunk 3]
```diff
@@ -796,7 +881,7 @@ class RecommendationSystemTest : public BackendTest {
     } else {
       bottomMLP = createMLP(mod, F, denseData, denseData->dims()[1],
                             bottomMLPIntermediateDims, embDim,
-                            numHiddenBottomMLPLayersOpt);
+                            numHiddenBottomMLPLayersOpt, true, true);
```
[the remark this answers is not in the document]
Comment (mlupon) [reply tests/unittests/RecommendationSystemTest.cpp]: I would like to keep the original behavior as default, and at the same time reduce the number of hyperparameters required to do the calls. Updated the code to only apply patches if the bias and weights factors are different as default.

**Comment by mlupon** [comment]
Pytorch CI failed during the compilation phase, but nothing has changed there. Maybe the Torch - Glow libraries are not in sync...

**Review by jfix71 — APPROVED** [review]
Thanks!

**Comment by jfix71** [comment]
@mlupon Apologies on the delay here, but there are now branch conflicts that will prevent me from importing/landing. Would you mind resolving these? Tag me once updated and I will import/land from there.

**Pull request closed without merging**